### PR TITLE
Rest service

### PR DESCRIPTION
**Pull request by NicolasDP**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
 /target
 **/*.rs.bk
+
+# IntelliJ project directory
+.idea
```
Comment (mzabaluev) [inline .gitignore]: I prefer to add my IDE-specific project files to `.git/info/ignore` rather than the version controlled `.gitignore`.

File: src/main.rs [hunk 2]
```diff
@@ -183,20 +201,31 @@ fn run() -> Result<(), Error> {
...
         tasks.task_create_with_inputs("block", move |r| {
-            block_task(blockchain, selection, clock, r, broadcast_sender)
+            block_task(
+                blockchain,
+                selection,
+                clock,
+                r,
+                broadcast_sender,
+                shared_stats,
+            )
```
Comment (mzabaluev) [inline src/main.rs]: It may improve maintainablility to group some or all of these these parameters into a struct at this point.

File: src/rest/server_service/error.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    Pkcs12Invalid(TlsError),
+    BindFailed(IoError),
+    ServerAlreadyStopped,
+    ServerStopTimeout,
+    ServerStopFailed,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let error_name = match *self {
```
Comment (mzabaluev) [inline src/rest/server_service/error.rs]: It's more conventional to use human-readable error messages in `Display`. Variant names are better suited for the `Debug` impl, which is macro-derived.

File: src/rest/server_service/error.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            Error::Pkcs12LoadFailed(_) => "Pkcs12LoadFailed",
+            Error::Pkcs12Invalid(_) => "Pkcs12Invalid",
+            Error::BindFailed(_) => "BindFailed",
+            Error::ServerAlreadyStopped => "ServerAlreadyStopped",
+            Error::ServerStopTimeout => "ServerStopTimeout ",
+            Error::ServerStopFailed => "ServerStopFailed ",
+        };
+        write!(f, "Server service error: {}", error_name)?;
```
Comment (mzabaluev) [inline src/rest/server_service/error.rs]: The choice of a prefix message should be left to a higher-level error handling code which knows it has this error type to represent. Just use a human-readable string describing the problem.

File: src/rest/server_service/error.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        if let Some(cause) = self.source() {
+            write!(f, "caused by {}", cause)?
```
Comment (mzabaluev) [inline src/rest/server_service/error.rs]: Don't do this here either, it's up to a top level error handler to walk the error chain if needed.

File: .gitignore (hunk 1 above)
Comment (CodeSandwich) [reply .gitignore]: Will fix in next MR

File: src/main.rs (hunk 2 above)
Comment (CodeSandwich) [reply src/main.rs]: I agree, 5 parameters are hard to follow. Should I refactor it now or wait with it until the demo? We will have to do quite some clean up then and developing fresh code is slightly easier with unpacked args.

File: src/rest/server_service/error.rs (hunk 3 above)
Comment (CodeSandwich) [reply src/rest/server_service/error.rs]: Display must be implemented for Error, it's a direct replacement for `Error::description()`. I don't think it must be actually displayed and understandable by end-user, at least not at this point.

File: src/rest/server_service/error.rs (hunk 4 above)
Comment (CodeSandwich) [reply src/rest/server_service/error.rs]: Will fix in next PR

File: src/main.rs (hunk 2 above)
Comment (mzabaluev) [reply src/main.rs]: Whichever works for you. I'd write even new code with easy to read structures that can be autocompleted.

File: src/rest/server_service/error.rs (hunk 3 above)
Comment (mzabaluev) [reply src/rest/server_service/error.rs]: Yes, but it should be human readable, no need to stuff things like enum variant names there.

Comment (CodeSandwich) [reply src/rest/server_service/error.rs]: You're right, that's the point of a description. I'll fix it in next PR.

File: src/rest/server_service/error.rs (hunk 5 above)
Comment (CodeSandwich) [reply src/rest/server_service/error.rs]: Will fix it in next PR

File: src/rest/server_service/error.rs (hunk 3 above)
Comment (mzabaluev) [reply src/rest/server_service/error.rs]: This PR is not getting anywhere without being rebased to resolve the conflicts, so you might as well fix it here.

Comment (CodeSandwich) [reply src/rest/server_service/error.rs]: This PR was already rebased and merged as a part of a different branch with additional functionalities. I would close or reuse this branch and this PR if I knew you were going to review it.

Comment (mzabaluev) [reply src/rest/server_service/error.rs]: OK, feel free to close this PR then.
